File: Default Components/Interfaces/coroutineScheduler.lua
local class = require('classlib')
local coroutineScheduler = class('coroutineScheduler')

function coroutineScheduler:__user_init()
  self.routines = {}
end

function coroutineScheduler:addRoutine(routine)
  assert(type(routine) == "function", "Must enter a function for a coroutine")

  table.insert(self.routines,
  {
    current = 0,
    target = 0,
    func = routine,
    routine = coroutine.create(routine)
  })
end

function coroutineScheduler:finishRoutine(routine)
  for _, value in pairs(self.routines) do
    if value.func == routine then
      local stauts
      repeat status = coroutine.resume(value.routine) until not status
      break
    end
  end
end

function coroutineScheduler:cancelRoutine(routine)
  for i, value in pairs(self.routines) do
    if value.func == routine then
      self.routines[i] = nil
      break
    end
  end
end

function coroutineScheduler:hasRoutine(routine)
  for _, value in pairs(self.routines) do
    if value.func == routine then return true end
  end
end

function coroutineScheduler:getStatus(routine)
  for _, value in pairs(self.routines) do
    if value.func == routine then
      return coroutine.status(value.routine)
    end
  end
end

function coroutineScheduler:clear()
  self.routines = {}
end

function coroutineScheduler:update(dt)
  for i, value in pairs(self.routines) do
    local routine = assert(value.routine, "No routine")
    local target = assert(value.target, "No target")
    if type(target) == "number" then
      local coroutineCurrent = assert(value.current, "No current")
      coroutineCurrent = coroutineCurrent + dt
      value.current = coroutineCurrent
      if coroutineCurrent < target then return end
    elseif classname(target) == "coroutineScheduler" then
      if coroutine.status(target.routine) == "running" then return end
    end

    local status, rval = coroutine.resume(routine)
    if status then
      if type(rval) == "number" or type(rval) == "nil" then
        value.target = rval or 0
        value.current = 0
      elseif classname(rval) == "coroutineScheduler" then
        value.target = rval
      end
    else
      self.routines[i] = nil
      if rval ~= "cannot resume dead coroutine" then
        print(rval)
      end
    end
  end
end

return coroutineScheduler

File: Javascript/Default Components/Interfaces/coroutineScheduler.js
define(function()
{
  class CoroutineScheduler
  {
    constructor()
    {
      this.routines = [];
    }

    addRoutine(func, args)
    {
      this.routines.push(
      {
        current : 0,
        target : 0,
        func : func(args),
        args : args
      });
    }

    finishRoutine(func)
    {
      for(var i = 0; i < this.routines.length; ++i)
      {
        var routine = this.routines[i];
        if(routine.func == func)
        {
          var result;
          do
          {
            result = func.next();
          }while(!result.done);
          break;
        }
      }
    }

    cancelRoutine(func)
    {
      for(var i = 0; i < this.routines.length; ++i)
      {
        var routine = this.routines[i];
        if(routine.func == func)
        {
          this.routines.splice(i, 1);
          break;
        }
      }
    }

    hasRoutine(func)
    {
      for(var i = 0; i < this.routines.length; ++i)
      {
        var routine = this.routines[i];
        if(routine.func == func)
        {
          return true;
        }
      }
      return false;
    }

    clear()
    {
      this.routines = [];
    }

    update(step)
    {
      for(var i = 0; i < this.routines.length; ++i)
      {
        var routine = this.routines[i];
        if(typeof routine.target == "number")
        {
          routine.current += step;
          if(routine.current < routine.target)
          {
            continue;
          }
        }
        else if(typeof routine.target == "object")
        {
          var rval = routine.target;
          if(rval.type == "frames")
          {
            ++routine.current;
            if(routine.current < rval.value)
            {
              continue;
            }
          }
          else if(rval.type == "time")
          {
            routine.current += step;
            if(routine.current < rval.value)
            {
              continue;
            }
          }
        }

        var result = routine.func.next(routine.args);
        if(result.done)
        {
          this.routines.splice(i--, 1);
        }
        else
        {
          routine.current = 0;
          if(result.value == null)
          {
            routine.target = 0;
          }
          else
          {
            routine.target = result.value;
          }
        }
      }
    }
  }

  return CoroutineScheduler;
});
